### ISSUE TITLE
[explicit-function-return-type] Allow implicit void functions

### ISSUE THREAD

**Issue by jorgegonzalez** [bradzacher/eslint-plugin-typescript#131]
It would be nice if this rule had an option to ignore errors for void functions not explicitly marked as such.

Currently:

```ts
// Invalid: Missing return type on function
function hello() {
	console.log('Hello');
}
```

After:

```json
{
	"rules": {
		"typescript/explicit-function-return-type": [
			"error",
			{
				"allowVoid": true
			}
		]
	}
}
```

```ts
// Valid
function hello() {
	console.log('Hello');
}

// Also valid
function something(props, text): void {
	props.text = text;
	return;
}
```

**Comment by armano2** [comment]
I think we should do it in different way, if there is no type specified, we should check if return type if consistent:
if all paths has or not have return like: `consistent-return` eslint rune does it.

it can be easily accomplished with [code path](https://eslint.org/docs/developer-guide/code-path-analysis/#code-path-analysis-details)

**Comment by bradzacher** [comment]
There's four cases you have to test on each path:
- no return at all
- `return;`
- `return undefined;`
- `x = () => void; return x();`
[repl](https://www.typescriptlang.org/play/index.html#src=function%20foo1()%3A%20void%20%7B%20%7D%0Afunction%20foo2()%3A%20void%20%7B%0A%20%20%20%20return%3B%0A%7D%0Afunction%20foo3()%3A%20void%20%7B%0A%20%20%20%20return%20undefined%3B%0A%7D%0Afunction%20foo4()%3A%20void%20%7B%0A%20%20%20%20return%20foo1()%3B%0A%7D)

All three are the same type-wise.
The last one is dubious, but is technically correct.
There has been talk of adding rules/features which statically analyse types, so this could potentially be a case for that...


The question is - what is the motive behind this rule?
IMO there are two motives:
1) documentation - it's easier to read the return type than it is to read the entire function to determine it's return type.
2) type soundness - you know that your return types are correctly inferred, and there are no invalid cases accidentally introduced. This is a great help esp if you don't have strict null checks turned on:

```TS
function foo(param : string) { // return type = string
  return 'string' + param;
}
foo('').length // runtime safe! 

// modify code

// strict return type = string | null
// non-strict return type = string
function foo(param : string) {
  if (!string) {
    return null;
  }

  return 'string' + param;
}
foo('').length // non-strict, compile-time safe, but breaks at runtime
```
[repl - use the options to turn strict nulls on/off](https://www.typescriptlang.org/play/index.html#src=function%20foo(param%3A%20string)%20%7B%0A%20%20%20%20return%20'string'%20%2B%20param%0A%7D%0Afoo('').length%0A%0Afunction%20foo2(param%3A%20string)%20%7B%0A%20%20%20%20if%20(!param)%20%7B%0A%20%20%20%20%20%20%20%20return%20null%0A%20%20%20%20%7D%0A%0A%20%20%20%20return%20'string'%20%2B%20param%0A%7D%0Afoo2('').length%0A)

So the question is - why strictly enforce return type on `void`?
Does disabling the check on `void` returns benefit the rule?

1) documentation
  - not explicitly stating the return type is void is ambiguous, so it breaks motivation 1.
  - "did their linter not run and they forgot to put the return type?"
2) type soundness
  - by not requiring voids, we still implicitly have the check for void (because if you add/remove a return value to/from a void function, the lint warning will throw because of the change in return type).
  - this doesn't break motivation 1, as it still ensures the same type soundness.

So the question is this - is the saving of 5-7 characters per void function and breaking of motivation (1) worth the code complexity that will be required to do function path and return type analysis?

**Comment by armano2** [comment]
why did you close this one?

**Comment by jorgegonzalez** [comment]
I understood that you didn't want to implement it ... and I've personally gotten used to explicitly typing `void`

**Comment by armano2** [comment]
currently we will have to add additional logic for this but, when type analysys will be live in parser we will have access to it directly. like accessing property. I will like to implement this feature, but not till we are going to update to parser v21.*

**Comment by jorgegonzalez** [comment]
In that case, forgive me for closing.
